src/librustc_mir/borrow_check/mod.rs

```diff
@@ -141,6 +142,40 @@ fn do_mir_borrowck<'a, 'gcx, 'tcx>(
         .as_local_node_id(def_id)
         .expect("do_mir_borrowck: non-local DefId");
 
+    // Gather the upvars of a closure, if any.
+    let upvars: Vec<_> = tcx.with_freevars(id, |freevars| {
```

Comment:
This really feels like information that could be encoded in the MIR, no? Well, I guess I can see an argument both ways.

Reply:
At minimum I would prefer if we can factor it out from *borrow check* into some kind of helper table that would be more re-usable.

Reply:
But I don't know that I want to block the PR on it. In general I feel like the borrow check code is kind of noodly and could use some extractions of this kind.